### ISSUE TITLE
udev does not seem to propagate changes (`KEYBOARD_KEY_*`)

### ISSUE THREAD

**Issue by ManDay** [systemd/systemd#20510]
Applying `KEYBOARD_KEY_*` environment attributes to an keyboard input device should have an effect, but it does not. Because it used to work with Gentoo's "eudev" fork, I'm confident that an assignment `KEYBOARD_KEY_01=capslock` should put `Capslock` on `Esc`. While the assignment shows up among others (supposedly originating from hwdb) in the database, it does not have any effect.

Remarks:
 - Observed on Gentoo's `sys-fs/udev` package
 - Curiously, I noticed that `systemd-hwdb` is missing from this package and is not installed; this could be related
 - I apologize if this is wrongly filed; here seemed more appropriate than on Gentoo's tracker

**Comment by yuwata** [comment]
Please try to generate debugging log of udevd. You can generate it by booting with `udev.log_level=debug` kernel command line.
Then, please check that the log contains something like the following:
```
event6: /usr/lib/udev/rules.d/60-evdev.rules:23 Importing properties from results of builtin command 'keyboard'
```

Also, please also show the result of `udevadm info /sys/class/input/eventX`.

**Comment by yuwata** [comment]
Also, please provide version of systemd.

**Comment by ManDay** [comment]
`udev.log_level=debug` does not have any effect and does not show in `sysctl` in linux-5.10.52.
The version of udev which should correspond to the systemd version is 248. This is the `udevadm` output you asked for:

```
P: /devices/platform/i8042/serio0/input/input3/event3
N: input/event3
L: 0
S: input/by-path/platform-i8042-serio-0-event-kbd
E: DEVPATH=/devices/platform/i8042/serio0/input/input3/event3
E: DEVNAME=/dev/input/event3
E: MAJOR=13
E: MINOR=67
E: SUBSYSTEM=input
E: USEC_INITIALIZED=1552268
E: KEYBOARD_KEY_86=wlan
E: KEYBOARD_KEY_a5=help
E: KEYBOARD_KEY_a6=setup
E: KEYBOARD_KEY_a7=battery
E: KEYBOARD_KEY_a9=switchvideomode
E: KEYBOARD_KEY_b2=www
E: KEYBOARD_KEY_b3=euro
E: KEYBOARD_KEY_b4=dollar
E: KEYBOARD_KEY_ce=brightnessup
E: KEYBOARD_KEY_d4=bluetooth
E: KEYBOARD_KEY_d5=wlan
E: KEYBOARD_KEY_d6=wlan
E: KEYBOARD_KEY_d7=bluetooth
E: KEYBOARD_KEY_d8=bluetooth
E: KEYBOARD_KEY_d9=bluetooth
E: KEYBOARD_KEY_ee=brightnessup
E: KEYBOARD_KEY_ef=brightnessdown
E: KEYBOARD_KEY_f1=f22
E: KEYBOARD_KEY_f2=f23
E: KEYBOARD_KEY_f3=prog2
E: KEYBOARD_KEY_f4=prog1
E: KEYBOARD_KEY_f5=presentation
E: KEYBOARD_KEY_f6=power
E: KEYBOARD_KEY_f8=fn
E: KEYBOARD_KEY_f9=prog1
E: KEYBOARD_KEY_84=bluetooth
E: KEYBOARD_KEY_92=media
E: ID_INPUT=1
E: ID_INPUT_KEY=1
E: ID_INPUT_KEYBOARD=1
E: ID_BUS=i8042
E: ID_SERIAL=noserial
E: ID_PATH=platform-i8042-serio-0
E: ID_PATH_TAG=platform-i8042-serio-0
E: LIBINPUT_DEVICE_GROUP=11/1/1:isa0060/serio0
E: KEYBOARD_KEY_01=capslock
E: KEYBOARD_KEY_3a=esc
E: DEVLINKS=/dev/input/by-path/platform-i8042-serio-0-event-kbd
```

**Comment by yuwata** [comment]
`udev.log_level=debug` should be set in the kernel command line. Not sysctl.

**Comment by ManDay** [comment]
That is what I did, but it does not have any effect and it does not show in `sysctl` when I inspect the commandline.

**Comment by yuwata** [comment]
Please find log by `journalctl`.

**Comment by yuwata** [comment]
BTW, in which file do you add `KEYBOARD_KEY_*`? The file name must be earlier than `60-evdev.rules` with alphanumerical order.

**Comment by ManDay** [comment]
Re `journalctl`, I don't run `journalctl`. Maybe I wasn't exactly clear, but this is only udev, I do not run the entire systemd stack. For what it's worth, there is no mention in my system log (i.e. `/var/log/messages` by `syslog-ng`), either.

`KEYBOARD_KEY_*` is added from `/etc/udev/rules.d/99escswap.rules`:

```
SUBSYSTEM=="input", ENV{ID_INPUT_KEYBOARD}=="1", ENV{KEYBOARD_KEY_01}="capslock", ENV{KEYBOARD_KEY_3a}="esc"
```

Can you please elaborate on the ordering, why it is important w.r.t. to `60-evdev.rules`?

**Comment by yuwata** [comment]
systemd-udevd processes udev rule files in alphanumerically. So, 99escswap.rules is processed after 60-evdev.rules. The environment variables `KEYBOARD_KEY_*` are used by udev's `keyboard` builtin command, and is called in 60-evdev.rules. Thus, the environment variables you specified in 99escswap.rules do not take any effects.

**Comment by yuwata** [comment]
> I did change the filename to 11-escswap.rules, but there was no change.

Did you reboot? Or, you need to call `udevadm control --reload` and `udevadm trigger --action=add /sys/class/input/eventX`.

**Comment by ManDay** [comment]
I did edit my above comment to include that ordering it in that other way is "even worse", in the sense that the my `KEYBOARD_KEY_*` entries do not appear. I think this is owing to the effect that the evdev rules are now processed later, *overwriting* my intended changes, so I think my original order is the correct one.

**Comment by yuwata** [comment]
`ID_INPUT_KEYBOARD=` is set by `input_id` builtin command, and is called in 60-input-id.rules. Please use another way to specify the device in your `11-escswap.rules`. How about the following?
```
SUBSYSTEM=="input", KERNEL=="event3", ENV{KEYBOARD_KEY_01}="capslock", ENV{KEYBOARD_KEY_3a}="esc"
```

**Comment by yuwata** [comment]
Or, call `input_id` in `11-escswap.rules`:
```
SUBSYSTEM=="input", ENV{ID_INPUT}=="", IMPORT{builtin}="input_id"
SUBSYSTEM=="input", ENV{ID_INPUT_KEYBOARD}=="1", ENV{KEYBOARD_KEY_01}="capslock", ENV{KEYBOARD_KEY_3a}="esc"
```

**Comment by ManDay** [comment]
Your suggestion in https://github.com/systemd/systemd/issues/20510#issuecomment-903307876 works, the swap is effected and both entries `KEYBOARD_KEY_*` appear in the database, as opposed to *my* rule named `11-escswap.rules`.

**Comment by yuwata** [comment]
OK. Let's close this. This is not a bug in udevd. But an issue in user configuration.

**Comment by ManDay** [comment]
Feel free to close this issue as PEBCAK. It would be extremely helpful if you could explain what is going on and why the same database output (both times with the `KEYBOARD_KEY_*` entries) has two different effects, depending on which context it comes from

- My attempt (which worked with the eudev fork): `99-escswap` and matched ` ENV{ID_INPUT_KEYBOARD}=="1"`
- Your solution: `11-escswap` and `KERNEL=="event3"`

**Comment by ManDay** [comment]
What I'm trying to ask is: If the correctness of the solution can not be determined by looking at the ouput of `udevadm info` because it looks the same both times, then how can it be determined - what else is involved and to be understood by the user?

**Comment by yuwata** [comment]
eudev used `RUN{builtin}+="keyboard"` while systemd-udevd uses `IMPORT{builtin}="keyboard"`. Commands specified in `RUN` are executed after all .rules files are processed, but `IMPORT` commands are executed immediately. That difference causes your issue.

**Comment by yuwata** [comment]
> What I'm trying to ask is: If the correctness of the solution can not be determined by looking at the ouput of `udevadm info` because it looks the same both times, then how can it be determined - what else is involved and to be understood by the user?

I do not have any general answer. But you can also use `udevadm test /sys/class/input/eventX` or `udevadm test-builtin keyboard /sys/class/input/eventX`.

**Comment by ManDay** [comment]
Thank you very much and please excuse the false report!

**Comment by yuwata** [comment]
> Thank you very much and please excuse the false report!

No problem.

**Comment by yuwata** [comment]
Note

> eudev used `RUN{builtin}+="keyboard"` while systemd-udevd uses `IMPORT{builtin}="keyboard"` in 60-eudev.rules. Commands specified in `RUN` are executed after all .rules files are processed, but `IMPORT` commands are executed immediately. That difference causes your issue.

Note that the change is introduced by 161bc525bbd7e3707cc71f0258f663bed9a354dd (systemd v245).

**Comment by matthijskooijman** [comment]
I have also been fiddling around with keyboard mapping from a udev rule (instead of hwdb) and was also having issues with getting the mappings applied. One thing that I would like to add for posterity (which might also have played a part in the original issue), is that it seems the rules in 60-evdev.rules call the hwdb and keyboard builtins in the same rule, which seems to mean that *if* hwdb does not match the device, the keyboard builtin is also not executed. So even when I put my custom rule before 60-evdev.rules, it was still not applied.

I fixed this by just explicitly running the keyboard builtin from my local 99-test.rules file (in a separate rule because it seems ENV entries are applied only after running the builtin, not in order of definition in the rule):

```
    SUBSYSTEM=="input", KERNEL=="event3", ENV{KEYBOARD_KEY_01}="capslock", ENV{KEYBOARD_KEY_3a}="esc"
    SUBSYSTEM=="input", KERNEL=="event3", IMPORT{builtin}="keyboard"
```

Note that I used 99 for the ordering - originally I used 00 which seemed to prevent the device from being marked as a keyboard (i.e. it was assigned to `/dev/input/by-id/usb-3412_7856-event-if00 ` instead of  `/dev/input/by-id/usb-3412_7856-event-kbd`). I have not further investigated what the implications and causes were (maybe the rule needed to be after some other rule than 60-evdev.rules).


Also, I've written down a bunch more stuff that I discovered about how hwdb and scancode mapping works in this post:  /dev/input/by-id/usb-3412_7856-event-kbd

**Comment by ManDay** [comment]
I'm encountering the same problem with an USB keyboard, and as before I find my ability to diagnose this to be very limited. I've used a HWDB file reading

```
evdev:atkbd:*
 KEYBOARD_KEY_3a=esc
 KEYBOARD_KEY_01=capslock
```

to successfully effect the remapping for the notebooks internal keyboard. I duplicated the section with a match for `evdev:name:*Keyboard*` to also apply it to an external USB keyboard. It's purely guesswork, but comparing "similar" device sections of `udevadm info -e` (the ones with `U: input`), the KV-pairs seem to be applied as needed. I get (three times, since the keyboard seems to create 3 distinct event nodes for some reason):

```
P: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.0/0003:1C4F:0002.0002/input/input5/event5
M: event5
R: 5,6,7
U: input
S: input/by-id/usb-SIGMACHIP_USB_Keyboard-event-kbd
S: input/by-path/pci-0000:00:14.0-usb-0:4:1.0-event-kbd
S: input/by-path/pci-0000:00:14.0-usbv2-0:4:1.0-event-kbd
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.0/0003:1C4F:0002.0002/input/input5/event5
E: SUBSYSTEM=input
…
E: KEYBOARD_KEY_01=capslock
E: KEYBOARD_KEY_3a=esc
```

And yet, there is no effect and `Esc` and `Capslock` remain unaltered. I'm of course reluctant to re-file this as a bug at this point; it would be great if you could advise!

**Comment by matthijskooijman** [comment]
@ManDay Maybe this post is helpful: https://unix.stackexchange.com/questions/397600/remapping-an-input-event-from-webcam-to-a-key-on-keyboard/734306#734306

I had originally intended to link this post from my previous comment, but rereading that comment just now I see I accidentally pasted the wrong thing there. I just updated my previous comment as well.

**Comment by ManDay** [comment]
@matthijskooijman Thanks to your link I found the problem, installing `evtest`: The USB keyboard seems to emit `0x70029` (not `0x1`) for `Esc` and `0x70039` (not `0x3a`) for `Capslock`. Given that the other values worked for the integrated keyboard, I suppose that was the catch; and I did not notice it because `showkeys -s` did, in fact, **not** show the *scan*codes but the *key*codes (i.e. `0x1` and `0x3a`), which just happend to coincide for the integrated keyboard.

Makes me wonder why `showkeys` is still around, seen how it fails to do the only thing it would be used for. But that's not an issue for here :slightly_smiling_face:
